src/core/mac/mac.cpp

```diff
@@ -794,26 +846,43 @@ void Mac::ReceiveDoneTask(Frame *aFrame, ThreadError aError)
 
     case sizeof(ShortAddress):
         aFrame->GetDstPanId(panid);
-        VerifyOrExit((panid == kShortAddrBroadcast || panid == mPanId) &&
-                     ((mRxOnWhenIdle && dstaddr.mShortAddress == kShortAddrBroadcast) ||
-                      dstaddr.mShortAddress == mShortAddress), ;);
+
+        if (!((panid == kShortAddrBroadcast || panid == mPanId) &&
+              ((mRxOnWhenIdle && dstaddr.mShortAddress == kShortAddrBroadcast) ||
+               dstaddr.mShortAddress == mShortAddress)))
+        {
+            error = kThreadError_Error;
+            ExitNow();
+        }
+
         break;
 
     case sizeof(ExtAddress):
         aFrame->GetDstPanId(panid);
-        VerifyOrExit(panid == mPanId &&
-                     memcmp(&dstaddr.mExtAddress, &mExtAddress, sizeof(dstaddr.mExtAddress)) == 0, ;);
+
+        if (panid != mPanId ||
+            memcmp(&dstaddr.mExtAddress, &mExtAddress, sizeof(dstaddr.mExtAddress)))
+        {
+            error = kThreadError_Error;
+            ExitNow();
+        }
+
         break;
     }
 
     // Security Processing
-    SuccessOrExit(ProcessReceiveSecurity(*aFrame, srcaddr, neighbor));
+    if (ProcessReceiveSecurity(*aFrame, srcaddr, neighbor) != kThreadError_None)
+    {
+        error = kThreadError_Security;
```

Comment:
This should work: `SuccessOrExit(error = ProcessReceiveSecurity(*aFrame, srcaddr, neighbor));`